src/Components/test/E2ETest/ServerRenderingTests/FormHandlingTests/FormWithParentBindingContextTest.cs

```diff
@@ -66,14 +68,422 @@ public void CanBindParameterToTheDefaultForm(bool suppressEnhancedNavigation)
+        var form = Browser.Exists(By.CssSelector("form"));
+        var formTarget = form.GetAttribute("action");
+        var actionValue = form.GetDomAttribute("action");
+        Assert.Equal(expectedTarget, formTarget);
+        Assert.Null(actionValue);
+
+        var name = Browser.Exists(By.CssSelector("""input[name="Model.Name"]"""));
+        name.SendKeys("John");
+        var email = Browser.Exists(By.CssSelector("""input[name="Model.Email"]"""));
+        email.SendKeys("john@example.com");
+        Browser.Click(By.CssSelector("""input[name="Model.IsPreferred"]"""));
+        Browser.Click(By.Id("send"));
+
+        Browser.Exists(By.Id("name")).Text.Contains("John");
+        Browser.Exists(By.Id("email")).Text.Contains("john@example.com");
+        Browser.Exists(By.Id("preferred")).Text.Contains("True");
+
+        if (!suppressEnhancedNavigation)
+        {
+            // Verify the same form element is still in the page
+            // We wouldn't be allowed to read the attribute if the element is stale
+            Assert.Equal(expectedTarget, form.GetAttribute("action"));
+        }
+    }
+
+    [Theory]
+    [InlineData(true)]
+    [InlineData(false)]
+    public void CanDisplayBindingErrorsComplexTypeToDefaultForm(bool suppressEnhancedNavigation)
+    {
+        var url = "forms/default-form-bound-complextype-parameter";
+        var expectedTarget = GetExpectedTarget(this, null, url);
+
+        if (suppressEnhancedNavigation)
+        {
+            GoTo("");
+            Browser.Equal("Hello", () => Browser.Exists(By.TagName("h1")).Text);
+            ((IJavaScriptExecutor)Browser).ExecuteScript("sessionStorage.setItem('suppress-enhanced-navigation', 'true')");
+        }
+
+        GoTo(url);
+
+        Browser.Exists(By.Id("ready"));
+        var form = Browser.Exists(By.CssSelector("form"));
+        var formTarget = form.GetAttribute("action");
+        var actionValue = form.GetDomAttribute("action");
+        Assert.Equal(expectedTarget, formTarget);
+        Assert.Null(actionValue);
+
+        var name = Browser.Exists(By.CssSelector("""input[name="Model.Name"]"""));
+        name.SendKeys("John");
+        var email = Browser.Exists(By.CssSelector("""input[name="Model.Email"]"""));
+        email.SendKeys("john@example.com");
+        Browser.Click(By.CssSelector("""input[name="Model.IsPreferred"]"""));
+        // Set value attribute to 'invalid' to trigger validation error
+        var isPreferred = Browser.Exists(By.CssSelector("""input[name="Model.IsPreferred"]"""));
+        ((IJavaScriptExecutor)Browser).ExecuteScript("arguments[0].setAttribute('value', 'invalid')", isPreferred);
+
+        Browser.Click(By.Id("send"));
+
+        Browser.Exists(By.CssSelector("li.validation-message")).Text.Contains("Could not convert value 'invalid' to 'System.Boolean'.");
+        Browser.Exists(By.CssSelector("div.validation-message")).Text.Contains("Could not convert value 'invalid' to 'System.Boolean'.");
```

Comment:
Just curious: when this surfaces the names of .NET types, are they definitely always low-level framework types, or can they be the names of application types? If it's the latter I'm surprised we'd be willing to disclose them in the UI by default. How does this compare with the kinds of messages that can be emitted by MVC?

Reply:
Sounds great!